enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,944 @@
+  consult the documentation.
+- Process would be unsymmetrically more difficult than upgrade, consider:
+  - Version A supports `v2`
+  - Version B supports `v1` and `v2`
+  - Version C supports `v1`
+  - To downgrade from version A to C
+    - Shutdown ostree commit A, boot commit B
+    - Instruct MicroShift to just downgrade data from `v2` to `v1`,
+      without running cluster (to not make migration too long)
+    - Persist metadata that version C will accept
+    - Shutdown ostree commit B, boot commit C
+    - MicroShift C would validate metadata to make sure it's compatible
+- Consequence of previous bullet - version metadata would need to go beyond
+  simple MicroShift version of X.Y.Z to not only tracking versions of all
+  resources, but perhaps versions of the embedded components as well.
+  It could be a case of internal implementation details that would support
+  newer and older behavior in newer version, but result in bugs when going back
+  to older version.
+
+### Alternative backup methods
+
+#### Copy-on-write
+
+Pros:
+- Underlying blocks are shared, so initially backup takes very little to no additional space
+Cons:
+- Not supported by all filesystems - requirement needs documenting
+
+#### etcdctl snapshot save/restore
+
+Pros
+- Database snapshot is much smaller than copy of database
+Cons:
+- Saved and restore etcd database doesn't contain whole history
+- Would require to ship `etcdctl` increasing footprint of MicroShift
+  which doesn't not happen at the moment
+
+#### Creating a tar file with data dir
+
+Pros:
+- backup in form of a single file
+Cons:
+- Without compression is weights as much as data dir
+
+### Symlinking live data to specific commit data
+
+Instead of making explicit backups, MicroShift could mimic ostree's way of
+managing the root filesystem: working directory would be a symlink to directory
+specific to ostree commit.
+However this posses some challenges in terms of allowing free "Z-stream traversal"
+and if healthy commits turned unhealthy due to problems with MicroShift's data
+there would not be any backup to restore from.
+
+### Pre run procedure being part of `microshift run`
+
+By putting pre run and run procedures together, there would be no way to
+specifically signal that pre run failed and block starting the cluster.
+Instead, systemd would restart the MicroShift service and possibly result in
+inconsistency in "history of commits".
+It is further amplified by the fact that greenboot runs health checks only once,
+on start, so it might means that also the data is in unexpected state due to
+attempts to migrate the data which might end up with different result than
+just after system start. This means that procedures should be robust and
+reproducible, not affected by what is current uptime of the system.
+
+## Infrastructure Needed [optional]
+
+N/A
+
+## Future Optimizations
+
+- Reimplement current greenboot health check for MicroShift in Go language as
```

Comment:
We will be continue development of greenboot to support additional cases, please feel free to open issues / create PR's in greenboot and we can discuss more on the subject.

Reply:
Sorry I misunderstood, it makes sense now.